### DISCUSSION TITLE
Contraction-compression stability issues

### DISCUSSION THREAD

**Discussion by benjaminlanthier** [jcmgray/quimb#200] [Q&A] [answered]
Hi @jcmgray, 

While working with the `compress_all()` function on my tensor networks (with a certain relative cutoff), I found that at a higher number of tensors, the results I get are wrong most of the time. I say that because I am supposed to get a positive number, but I sometimes end up with a negative value, which made me think of either a bad `cutoff` for the compression function or a precision/stability problem. I tried minimizing the `cutoff` and I still got wrong results, so I was now looking for ways to increase the stability of my contractions/compression.

While looking for some solutions, I saw the functions `distribute_exponent()` and `equalize_norms()` and together, they seem to be a good tool to bring values to a certain acceptable level. That's what I could find in the documentation, but I was wondering if there are any other tools that are available to help with stability issues ?

**Comment by jcmgray** [comment, marked as answer]
Hi @benjaminlanthier, sorry this should probably have been noted in the docs but the `compress_all` function is really the most naive and basic way to compress a tensor network and not really meant to be used for any real applications (see also #153). If you need a relatively cheap way to compress an arbitrary network, I'd use `compress_all_simple`. Or beyond that think about some manual methods suiting your geometry and tensor entries.

The equalize_norms functionality helps with problems of numerical range (e.g. values >1e300) but fundamentally you will need some type of gauging (i.e. taking into account more than just local tensors) to get good accuracy beyond very simple TNs.

**Reply by benjaminlanthier** [reply]
Okay, thanks for the answer! May I ask what is the difference between `compress_all()` and `compress_all_simple()` ?

**Reply by jcmgray** [reply]
compress_all just iterates though all the bonds and compresses them locally. compress_all_simple on the other hand runs essentially simple update on the whole tensor network, then simultaneously truncates everywhere using the ‘simple’ gauges. It’s also called “super orthogonalization”.

**Reply by benjaminlanthier** [reply]
Perfect, thanks for the explanation!
